feature/validation

ncocquio opened the pull request:
#8

Commit:
WIP: Validation of controls

Commit:
Add network error

Commit:
Add email, duration and select validators
- to employee create/edit
- to project create/edit
- to contract create/edit

Commit:
Add input form restrictions

kelvinlouis commented on src/components/ProjectModelForm/ProjectCreateForm.tsx:
```diff
@@ -51,14 +52,19 @@ export const ProjectCreateForm: Component<ProjectFormProps> = ({ state, actions
               name={ftePercentage.name}
               value={ftePercentage.value}
+              min={1}
+              max={9223372036854775807}
```
Please use a constant

kelvinlouis commented on src/components/ProjectModelForm/ProjectCreateForm.tsx:
```diff
@@ -51,14 +52,19 @@ export const ProjectCreateForm: Component<ProjectFormProps> = ({ state, actions
               type="text"
               onInputChange={formActions.updateValue}
             />
+            {name.errors == null && <FormHint label={String.fromCharCode(160)} />}
```
Remove all placeholders and use min-height in css instead

kelvinlouis commented on src/components/FormHint/FormHint.tsx:
```diff
@@ -0,0 +1,24 @@
+    className = `${className} is-${theme}`;
+  }
+
+  return (
```
Please put everything on one line

kelvinlouis commented on src/components/EmployeeModalForm/EmployeeCreateForm.tsx:
```diff
@@ -49,25 +49,35 @@ export const EmployeeCreateForm: Component<Props> = ({ state, actions }) => {
             <FormInput
               name={firstName.name}
               value={firstName.value}
+              maxLength={50}
```
Pass errors to `FormInput`, `FormSelect`, `DatePicker`, `EmployeeSelect` and `FormCheckbox` and mark the control with `is-danger` if it contains errors.

kelvinlouis replied:
Add constants for 50 => `INPUT_MIN_LENGTH_SHORT = 50`

kelvinlouis commented on src/components/ContractForm/ContractForm.tsx:
```diff
@@ -45,19 +48,22 @@ export const ContractForm: Component<Props> = ({ state, actions, key }) => {
             type="number"
             suffix="fas fa-percent"
-            min={0}
+            min={1}
```
Add constant `PENSUM_VALUE_MIN` resp. `PENSUM_VALUE_MAX`.

kelvinlouis commented on src/components/EmployeeModalForm/EmployeeCreateForm.tsx:
```diff
@@ -49,25 +49,35 @@ export const EmployeeCreateForm: Component<Props> = ({ state, actions }) => {
             <FormInput
               name={lastName.name}
               value={lastName.value}
+              maxLength={50}
```
Reuse the component from above.

kelvinlouis commented on src/components/EmployeeModalForm/EmployeeCreateForm.tsx:
```diff
@@ -49,25 +49,35 @@ export const EmployeeCreateForm: Component<Props> = ({ state, actions }) => {
             <FormInput
               name={emailAddress.name}
               value={emailAddress.value}
+              maxLength={120}
```
Add constants for 120 => `INPUT_MIN_LENGTH_MEDIUM = 120`

kelvinlouis commented on src/components/EmployeeModalForm/EmployeeCreateForm.tsx:
```diff
@@ -76,14 +86,17 @@ export const EmployeeCreateForm: Component<Props> = ({ state, actions }) => {
               labelText="Active"
               onInputChange={formActions.updateValue}
             />
+            {active.errors == null && <FormHint label={"Inactive: No login possible"} />}
```
remove the braces

kelvinlouis commented on src/components/EmployeeModalForm/EmployeeEditForm.tsx:
```diff
@@ -50,44 +51,57 @@ export const EmployeeEditForm: Component<Props> = ({ state, actions }) => {
             <FormInput
               name={firstName.name}
               value={firstName.value}
+              maxLength={50}
```
Reuse constant

kelvinlouis commented on src/components/EmployeeModalForm/EmployeeEditForm.tsx:
```diff
@@ -50,44 +51,57 @@ export const EmployeeEditForm: Component<Props> = ({ state, actions }) => {
             <FormInput
               name={emailAddress.name}
               value={emailAddress.value}
+              maxLength={120}
```
Reuse constant

kelvinlouis commented on src/components/ProjectModelForm/ProjectCreateForm.tsx:
```diff
@@ -51,14 +52,19 @@ export const ProjectCreateForm: Component<ProjectFormProps> = ({ state, actions
             <FormInput
               name={name.name}
               value={name.value}
+              maxLength={50}
```
Reuse constant

kelvinlouis commented on src/components/ProjectModelForm/ProjectCreateForm.tsx:
```diff
@@ -51,14 +52,19 @@ export const ProjectCreateForm: Component<ProjectFormProps> = ({ state, actions
             <FormInput
               name={ftePercentage.name}
               value={ftePercentage.value}
+              min={1}
```
Reuse constant

kelvinlouis commented on src/state/form/employee-form.state.ts:
```diff
@@ -23,22 +24,38 @@ export const initEmployeeForm: () => EmployeeFormState = () => ({
       name: 'role',
       value: null,
+      validators: [
+        selectValidator("Please select"),
```
Single quote

kelvinlouis commented on src/state/form/validators.ts:
```diff
@@ -0,0 +1,53 @@
+
+export const selectValidator = (defaultValue: string) => (control: FormControl<string>, state: BaseForm): FormErrors => {
+
+  if (control.value != null && control.value === defaultValue) {
```
Maybe use `value` for the placeholder option. This way you could ignore `defaultValue` as a parameter.

Commit:
Fixed flaws from review
- Add constants
- Propagate error to form fields
- Cleanup select validation
- CSS hack for hints

Commit:
Move property one level up

kelvinlouis commented on src/components/ContractForm/ContractForm.tsx:
```diff
@@ -35,6 +37,8 @@ export const ContractForm: Component<Props> = ({ state, actions, key }) => {
             max={endDate.value}
             onInputChange={updateContractFormValue(key, formActions)}
           />
+          {startDate.errors == null && <FormHint label={String.fromCharCode(160)} />}
```
You shall remove

kelvinlouis replied:
Make sure you pass the errors object to the FormControl

kelvinlouis commented on src/components/ContractForm/ContractForm.tsx:
```diff
@@ -45,19 +49,22 @@ export const ContractForm: Component<Props> = ({ state, actions, key }) => {
             min={startDate.value}
             onInputChange={updateContractFormValue(key, formActions)}
           />
+          {endDate.errors == null && <FormHint label={String.fromCharCode(160)} />}
```
You shall remove

kelvinlouis replied:
Add the errors to the form control

kelvinlouis commented on src/components/ContractForm/ContractForm.tsx:
```diff
@@ -45,19 +49,22 @@ export const ContractForm: Component<Props> = ({ state, actions, key }) => {
+            max={CONTRACT_PENSUM_VALUE_MAX}
             onInputChange={updateContractFormValue(key, formActions)}
           />
+          {pensumPercentage.errors == null && <FormHint label={String.fromCharCode(160)} />}
```
You shall remove

kelvinlouis commented on src/components/ContractForm/ContractForm.tsx:
```diff
@@ -45,19 +49,22 @@ export const ContractForm: Component<Props> = ({ state, actions, key }) => {
             suffix="fas fa-percent"
-            min={0}
-            max={100}
+            min={CONTRACT_PENSUM_VALUE_MIN}
```
you might want to add errors here too.

ncocquio replied:
We eliminated the textLengthValidator in favor of HTML input limits.
Spinner has limits set, so a validator is obsolete.

kelvinlouis commented on src/components/EmployeeModalForm/EmployeeCreateForm.tsx:
```diff
@@ -49,41 +50,54 @@ export const EmployeeCreateForm: Component<Props> = ({ state, actions }) => {
               value={emailAddress.value}
+              maxLength={INPUT_LENGTH_LONG_MAX}
               type="email"
+              hasError={emailAddress.errors != null && (emailAddress.errors.required || emailAddress.errors.email)}
```
emailAddress.errors != null should suffice.

kelvinlouis commented on src/components/EmployeeModalForm/EmployeeCreateForm.tsx:
```diff
@@ -49,41 +50,54 @@ export const EmployeeCreateForm: Component<Props> = ({ state, actions }) => {
               name={password.name}
               value={password.value}
               type="password"
+              hasError={password.errors != null && password.errors.required}
```
password.errors != null should suffice

kelvinlouis commented on src/components/EmployeeModalForm/EmployeeCreateForm.tsx:
```diff
@@ -49,41 +50,54 @@ export const EmployeeCreateForm: Component<Props> = ({ state, actions }) => {
               value={lastName.value}
+              maxLength={INPUT_LENGTH_SHORT_MAX}
               type="text"
+              hasError={lastName.errors != null && lastName.errors.required}
```
i would change `hasError` to `errors` and pass it the entire `errors` object. The `errors` object will be undefined, if there are no errors.

kelvinlouis commented on src/components/EmployeeModalForm/EmployeeCreateForm.tsx:
```diff
@@ -49,41 +50,54 @@ export const EmployeeCreateForm: Component<Props> = ({ state, actions }) => {
+            {password.errors != null && password.errors.required && <FormHint theme="danger" label="Password is required" />}
           </FormField>
-          <FormField labelText="Status" required={true} hint="Inactive: No login possible">
+          <FormField labelText="Status" required={true}>
```
You might want to add the errors to the FormCheckbox too.

ncocquio replied:
Checkbox has no styling options

kelvinlouis commented on src/components/EmployeeModalForm/EmployeeEditForm.tsx:
```diff
@@ -50,44 +52,57 @@ export const EmployeeEditForm: Component<Props> = ({ state, actions }) => {
+            {emailAddress.errors != null && emailAddress.errors.email && <FormHint theme="danger" label="Email invalid" />}
           </FormField>
-          <FormField labelText="Status" required={true} hint="Inactive: No login possible">
+          <FormField labelText="Status" required={true}>
```
Add errors here too.

kelvinlouis commented on src/components/FormControlProps.ts:
```diff
@@ -5,4 +5,5 @@ export interface FormControlProps<T> extends FormControl<T> {
   placeholder?: string;
   disabled?: boolean;
   isLoading?: boolean;
+  hasError?: boolean;
```
pass it the FormErrors type:

`errors: FormErrors`

kelvinlouis commented on src/components/FormInput/FormInput.tsx:
```diff
@@ -20,7 +22,9 @@ const FormInputSuffix: Component<FormInputSuffixProps> = ({ suffix }) => (
-  const inputClassName = props.isLoading ? 'input is-loading' : 'input';
+  let inputClassName = 'input';
+  inputClassName = props.isLoading ? inputClassName + ' is-loading' : inputClassName;
+  inputClassName = props.hasError ? inputClassName + ' is-danger' : inputClassName;
```
```
inputClassName = props.errors != null ? inputClassName + ' is-danger' : inputClassName;
```

kelvinlouis commented on src/components/FormSelect/FormSelect.tsx:
```diff
@@ -52,6 +52,8 @@ export const FormSelect: Component<FormSelectProps<any, any>> = (props) => {
     value = o => o != null ? props.valueMapper!(o) : null;
   }
 
+  const selectClassName = props.hasError ? 'select is-danger' : 'select';
```
const selectClassName = props.errors != null ? 'select is-danger' : 'select';

Commit:
Simplify error handling in form elements
- Add bulma-checkradio

Commit:
Merged develop

Commit:
Format date in ProjectListItem

Commit:
fixed linting errors
